### PR TITLE
feat(youtube-scrape): Add yt scrape option

### PR DESCRIPTION
**Pull request by KurokuTetsuya**

### CODE REVIEW

**Comment by KurokuTetsuya** [comment]
This will solve #224

**Comment by KurokuTetsuya** [comment]
image
Still have some problem with playlist extraction.

File: .env.example [hunk 1]
```diff
@@ -20,4 +20,5 @@ CONFIG_STATUS_ACTIVITY="music on {guildsCount} guilds!" # What status does the b
 CONFIG_FETCH_ALL_USERS=no # Whether to fetch all users on startup (warning: increased memory usage especially for bots with large users) (default: no) (note: You need to enable privileged GUILD_MEMBERS Intents on the developer page)
 CONFIG_SELECT_TIMEOUT=20 # Select song timeout when searching songs (in seconds) (default: 20 seconds)
 CONFIG_DISABLE_SONG_SELECTION=no # Whether to disable selecting songs on play command, this will automatically selects the first song from search results (default: no)
-CONFIG_SEARCH_MAX_RESULTS=12 # Maximum songs to list in song selection (default: 12, max: 12)
\ No newline at end of file
+CONFIG_SEARCH_MAX_RESULTS=12 # Maximum songs to list in song selection (default: 12, max: 12)
+CONFIG_YOUTUBE_SCRAPE=yes # Whether to scrape youtube instead using api. You may not fill the API config if you use this.
```
Comment (Hazmi35) [inline .env.example:24]:
```suggestion
CONFIG_YOUTUBE_SCRAPE=yes # Whether to scrape youtube instead using API.
```
This description should be moved to YT_API_KEY's comment

File: .env.schema [hunk 2]
```diff
@@ -18,3 +18,4 @@ CONFIG_FETCH_ALL_USERS=
 CONFIG_SELECT_TIMEOUT=
 CONFIG_DISABLE_SONG_SELECTION=
 CONFIG_SEARCH_MAX_RESULTS=
+CONFIG_YOUTUBE_SCRAPE=
```
Comment (Hazmi35) [inline .env.schema:21]: Please add this environment variable in app.json too

File: src/commands/PlayCommand.ts [hunk 3]
```diff
@@ -36,6 +38,54 @@ export class PlayCommand extends BaseCommand {
             );
         }
 
+        if (this.client.config.youtubeScrape) {
```
Comment (Hazmi35) [inline src/commands/PlayCommand.ts:41]: This should be moved to the Jukebox client, but by this, you need to make the interface the same as YoutubeAPI

File: src/config.ts [hunk 4]
```diff
@@ -19,7 +19,9 @@ export const fetchAllUsers = process.env.CONFIG_FETCH_ALL_USERS === "yes";
 export const selectTimeout = Number(process.env.CONFIG_SELECT_TIMEOUT) * 1000 || 20 * 1000;
 export const disableSongSelection = process.env.CONFIG_DISABLE_SONG_SELECTION === "yes";
 export const searchMaxResults = Number(process.env.CONFIG_SEARCH_MAX_RESULTS) || 12;
+export const youtubeScrape = process.env.CONFIG_YOUTUBE_SCRAPE === "yes";
 
 if (searchMaxResults < 1) throw new Error("CONFIG_SEARCH_MAX_RESULTS cannot be smaller than 1");
 if (searchMaxResults > 12) throw new Error("CONFIG_SEARCH_MAX_RESULTS cannot be higher than 12");
 if (totalShards !== "auto" && isNaN(totalShards as unknown as number)) throw new Error("CONFIG_TOTALSHARDS must be a number or \"auto\"");
+if (youtubeScrape && process.env.YT_API_KEY) console.warn("[WARN] You may not fill YT_API_KEY because you enable youtube scrapping which doesn't require API key.");
```
Comment (Hazmi35) [inline src/config.ts:27]: Is this necessary? I think not

File: src/utils/YoutubeScrape/index.ts [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+    private extractData(video: any): VideoInfo {
+        const title = video.title.runs[0].text;
+        const author = video.shortBylineText?.runs[0].text;
+        const duration = video.lengthText ? this.durationToMs(video.lengthText.simpleText) : null;
+        return new VideoInfo(video.videoId, title, author, duration);
+    }
+
+    private durationToMs(duration: string): number {
+        return duration.split(/[:.]/).map(Number).reduce((acc, curr) => curr + acc * 60) * 1000;
+    }
```
Comment (Hazmi35) [inline src/utils/YoutubeScrape/index.ts:80]: Can you use ISO8601 Duration instead?

File: src/utils/YoutubeScrape/structures/Video.ts [hunk 6]
```diff
@@ -0,0 +1,10 @@
+export class VideoInfo {
+    public readonly url: string;
+    public readonly thumbnailURL: string;
+    public constructor(public readonly id: string, public readonly title: string, public readonly author: string, public readonly duration: number|null) {
+        this.url = `https://www.youtube.com/watch?v=${this.id}`;
+        this.thumbnailURL = `https://i.ytimg.com/vi/${this.id}/hqdefault.jpg`;
+        // i.ytimg.com may not available for some videos (the quality)
+        // You can do manual eliminate by using http request and check the response status
```
Comment (Hazmi35) [inline src/utils/YoutubeScrape/structures/Video.ts:8]: Hmm, I'll try to find a solution to this

**Review by Hazmi35 — CHANGES_REQUESTED** [review]

File: .env.example [hunk 7]
```diff
@@ -1,5 +1,5 @@
 DISCORD_TOKEN=Njk1NjY0ODU4NDA1Nzk3ODg4.dQw4w9WgXcQ.TmV2ZXIgZ29ubmEgZ2l2ZSB5b3UgdXA # Your Discord bot's token
-YT_API_KEY=IAzSjcSjaIEWakos-_AKfaKejiwdjiASJIAsDs # Your Youtube API key 
+YT_API_KEY=IAzSjcSjaIEWakos-_AKfaKejiwdjiASJIAsDs # Your Youtube API key. You may not fill the API config if you use CONFIG_YOUTUBE_SCRAPE.
```
Comment (Hazmi35) [inline .env.example:2]:
```suggestion
YT_API_KEY=IAzSjcSjaIEWakos-_AKfaKejiwdjiASJIAsDs # Your Youtube API key. You don't need to fill this if you use CONFIG_YOUTUBE_SCRAPE.
```

File: app.json [hunk 8]
```diff
@@ -7,7 +7,7 @@
             "description": "Your Discord bot's token"
         },
         "YT_API_KEY": {
-            "description": "Youtube API Key for querying the Youtube API for video data."
+            "description": "Youtube API Key for querying the Youtube API for video data. You may not fill the API config if you use CONFIG_YOUTUBE_SCRAPE."
```
Comment (Hazmi35) [inline app.json:10]:
```suggestion
            "description": "Your Youtube API key. You don't need to fill this if you use CONFIG_YOUTUBE_SCRAPE."
```

File: app.json [hunk 9]
```diff
@@ -75,6 +75,9 @@
         "CONFIG_SEARCH_MAX_RESULTS": {
             "description": "Maximum songs to list in song selection (default: 12, max: 12)",
             "required": false
+        },
+        "CONFIG_YOUTUBE_SCRAPE": {
+            "description": "Whether to scrape youtube instead using API."
```
Comment (Hazmi35) [inline app.json:80]:
```suggestion
            "description": "Whether to scrape youtube instead using API.",
            "required": false
```

File: src/commands/PlayCommand.ts (hunk 3 above)
Comment (Hazmi35) [reply src/commands/PlayCommand.ts:40]: Why is this resolved?

File: src/structures/Jukebox.ts [hunk 10]
```diff
@@ -9,11 +9,13 @@ import { YoutubeAPI } from "../utils/YoutubeAPI";
 
 // Extends DiscordJS Structures
 import "./Guild";
+import { YoutubeScrape } from "../utils/YoutubeScrape";
 
 export class Jukebox extends Client {
     public readonly config = config;
     public readonly logger = createLogger("main", config.debug);
     public readonly youtube = new YoutubeAPI(process.env.YT_API_KEY!);
+    public readonly youtubeScrapper = new YoutubeScrape(this);
```
Comment (Hazmi35) [inline src/structures/Jukebox.ts:18]: I think you should make merge this to Jukebox#youtube

**Review by Hazmi35 — CHANGES_REQUESTED** [review]

**Review by mzrtamp — APPROVED** [review]

**Comment by mzrtamp** [comment]
🙁

**Comment by mzrtamp** [comment]
[`ZenForks:feat/youtube-scrape`](https://github.com/ZenForks/jukebox/pull/4) why could this thing be skipped?

**Pull request closed without merging**